minerl/env/malmo.py

```diff
@@ -508,53 +509,52 @@ def launch(self):
                     raise RuntimeError("Malmo failed to start the MalmoEnv server! Check the logs from the Minecraft process.");
                 self._logger.info("Minecraft process ready")
                  
-
                 if not port == self._port:
                     self._logger.warning("Tried to launch Minecraft on port {} but that port was taken, instead Minecraft is using port {}.".format(port, self.port))
+
                 # supress entire output, otherwise the subprocess will block
                 # NB! there will be still logs under Malmo/Minecraft/run/logs
                 # FNULL = open(os.devnull, 'w')
                 # launch a logger process
-                def log_to_file(logdir):
-                    if not os.path.exists(os.path.join(logdir, 'logs')):
-                            os.makedirs((os.path.join(logdir, 'logs')))
+                def malmo_java_stdout_log_to_file(logdir):
+                    os.makedirs(os.path.join(logdir, 'logs'), exist_ok=True)
+                    log_path = os.path.join(logdir, 'logs',
+                                            'mc_{}.log'.format(self._target_port))
+                    self._logger.info("Logging Minecraft+Malmo java instance to {}".format(log_path))
+
+                    MB = 1024**2
+                    direct_format = logging.Formatter('%(message)s')
+                    log_handler = logging.handlers.RotatingFileHandler(
+                        log_path, mode='w', maxBytes=5 * MB,
+                        backupCount=2, encoding=None, delay=0)
+                    log_handler.setFormatter(direct_format)
+                    java_stdout_to_file_logger = logging.getLogger('malmo_java_stdout')
+                    java_stdout_to_file_logger.addHandler(log_handler)
+                    java_stdout_to_file_logger.setLevel(logging.INFO)
+
+                    mine_log_encoding = locale.getpreferredencoding(False)
 
-                    file_path = os.path.join(logdir, 'logs', 'mc_{}.log'.format(self._target_port - 9000))
+                    while self.running:
+                        line = self.minecraft_process.stdout.readline()
+                        if not line:
+                            break
 
-                    logger.info("Logging output of Minecraft to {}".format(file_path))
+                        linestr = line.decode(mine_log_encoding)
+                        linestr = "\n".join(linestr.split("\n")[:-1])
+                        if 'STDERR' in linestr or 'ERROR' in linestr:
+                            # Opportune place to suppress harmless MC errors.
+                            if not ('hitResult' in linestr):
+                                self._logger.error(linestr)
+                        elif 'LOGTOPY' in linestr:
```

Comment:
When does "LOGTOPY" appear in a line? These are from parts of Malmo, presumably?

Reply:
I don't know what this is. Yes, this is probably a Malmo thing.

Reply:
Oh, it's Log to py" that makes sense